File: backend/src/main/java/com/projectflow/service/ProjectService.java
package com.projectflow.service;

import com.projectflow.dto.request.ProjectRequest;
import com.projectflow.dto.response.ApiResponse;
import com.projectflow.entity.User;
import org.springframework.data.domain.Pageable;

public interface ProjectService {
    ApiResponse.ProjectResponse createProject(User user, ProjectRequest.Create request);
    ApiResponse.ProjectResponse getProject(User user, Long projectId);
    ApiResponse.PageResponse<ApiResponse.ProjectResponse> getUserProjects(User user, String search, Pageable pageable);
    ApiResponse.ProjectResponse updateProject(User user, Long projectId, ProjectRequest.Update request);
    void deleteProject(User user, Long projectId);
    ApiResponse.ProjectResponse inviteMember(User user, Long projectId, ProjectRequest.InviteMember request);
    void removeMember(User user, Long projectId, Long memberId);
    ApiResponse.ProjectResponse updateMemberRole(User user, Long projectId, Long memberId, ProjectRequest.UpdateMemberRole request);
}

File: frontend/src/app/core/services/project.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import {
  Project, CreateProjectRequest, UpdateProjectRequest,
  PageResponse, KanbanBoard
} from '../../shared/models';

@Injectable({ providedIn: 'root' })
export class ProjectService {
  private http = inject(HttpClient);
  private apiUrl = `${environment.apiUrl}/projects`;

  getProjects(search?: string, page = 0, size = 20): Observable<PageResponse<Project>> {
    let params = new HttpParams().set('page', page).set('size', size);
    if (search) params = params.set('search', search);
    return this.http.get<PageResponse<Project>>(this.apiUrl, { params });
  }

  getProject(id: number): Observable<Project> {
    return this.http.get<Project>(`${this.apiUrl}/${id}`);
  }

  createProject(request: CreateProjectRequest): Observable<Project> {
    return this.http.post<Project>(this.apiUrl, request);
  }

  updateProject(id: number, request: UpdateProjectRequest): Observable<Project> {
    return this.http.put<Project>(`${this.apiUrl}/${id}`, request);
  }

  deleteProject(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  inviteMember(projectId: number, email: string, role = 'MEMBER'): Observable<Project> {
    return this.http.post<Project>(`${this.apiUrl}/${projectId}/members`, { email, role });
  }

  removeMember(projectId: number, memberId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${projectId}/members/${memberId}`);
  }

  updateMemberRole(projectId: number, memberId: number, role: string): Observable<Project> {
    return this.http.patch<Project>(`${this.apiUrl}/${projectId}/members/${memberId}/role`, { role });
  }
}
